### PR TITLE
refactor(es/minifier): Respect `toplevel` and `module` options

### PR DESCRIPTION
**Pull request by Austaras**
**Description:**

**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

File: crates/swc_ecma_minifier/tests/benches-full/echarts.js [hunk 1]
```diff
@@ -15415,27 +15415,27 @@
                                 }
                             ]);
                             for(var i = 0; i < lastLevelTicks.length - 1; i++){
-                                var approxInterval1, approxInterval2, approxInterval3, startTick = lastLevelTicks[i].value, endTick = lastLevelTicks[i + 1].value;
+                                var approxInterval1, startTick = lastLevelTicks[i].value, endTick = lastLevelTicks[i + 1].value;
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/benches-full/echarts.js:15418]: No observable effect

File: crates/swc_ecma_minifier/tests/fixture/issues/2044/full/output.js [hunk 2]
```diff
@@ -15,14 +16,4 @@ Object.defineProperty(void 0, "__esModule", {
...
-}(), (void 0).default = void 0, function(exports) {
-    exports.default = void 0, (void 0).exports = exports.default;
-}(), function(exports) {
-    exports.default = void 0, (void 0).exports = exports.default;
-}(), function(exports) {
-    exports.default = void 0, exports.default = String;
-}(), function(exports) {
-    exports.default = void 0, exports.default = String;
-}(), (void 0).__esModule = !0, (void 0).__esModule = !0, (void 0).__esModule = !0;
+}(), exports.default = null, (void 0).exports = exports.default, (void 0).default = void 0, exports1.default = void 0, (void 0).exports = exports1.default, exports2.default = void 0, (void 0).exports = exports2.default, exports3.default = void 0, exports3.default = String, exports4.default = void 0, exports4.default = String, (void 0).__esModule = !0, (void 0).__esModule = !0, (void 0).__esModule = !0;
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/fixture/issues/2044/full/output.js:19]: Intended.

File: crates/swc_ecma_minifier/tests/fixture/issues/6509/output.js [hunk 3]
```diff
@@ -12,6 +12,7 @@ export var modifiers = {
                 0
             ];
             else {
+                var offset1, popperOffsets, referenceOffsets;
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/fixture/issues/6509/output.js:15]: This test case doesn't enable `unused` or `side_effect`.

File: crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2436_2/output.js [hunk 4]
```diff
@@ -1,7 +1,8 @@
...
-    })(o)
-);
+var o = {
+    a: 1,
+    b: 2
+};
+console.log((o.a = 3, {
+    x: o.a,
+    y: o.b
+}));
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2436_2/output.js:8]: `sequences` take effect without being configured.

File: crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2436_3/output.js [hunk 5]
```diff
@@ -1,7 +1,11 @@
...
+    a: 1,
+    b: 2
+};
+console.log((c = o, o = {
+    a: 3,
+    b: 4
+}, {
+    x: c.a,
+    y: c.b
+}));
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2436_3/output.js:11]: Same as above.

File: crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2436_5/output.js [hunk 6]
```diff
@@ -1,5 +1,8 @@
...
-    })({ a: 1, b: 2 })
-);
+var o;
+console.log({
+    x: (o = {
+        a: 1,
+        b: 2
+    }).a,
+    y: o.b
+});
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2436_5/output.js:8]: terser cannot handle duplicate idents.

File: crates/swc_ecma_minifier/tests/terser/compress/dead_code/dead_code_const_declaration/output.js [hunk 7]
```diff
@@ -1,3 +1,3 @@
 var unused;
 const CONST_FOO = !1;
-var moo, bar;
+var moo;
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/terser/compress/dead_code/dead_code_const_declaration/output.js:3]: https://github.com/terser/terser/issues/1274

File: crates/swc_ecma_minifier/tests/terser/compress/functions/drop_lone_use_strict_arrows_1/output.js [hunk 8]
```diff
@@ -1,6 +1,3 @@
-var f0 = () => 0;
-var f1 = () => {};
-var f2 = () => {
-    "use strict";
-    var f3 = () => {};
-};
+var f0 = ()=>0;
+var f1 = ()=>{};
+var f2 = ()=>{};
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/terser/compress/functions/drop_lone_use_strict_arrows_1/output.js:3]: unused pass is enabled when `side_effects` set to true in swc.

File: crates/swc_ecma_minifier/tests/terser/compress/functions/loop_init_arg/output.js [hunk 9]
```diff
@@ -1,4 +1,6 @@
 var a = "PASS";
-for (var k in "12") (b = void 0), (b >>= 1) && (a = "FAIL"), (b = 2);
-var b;
+for(var k in "12"){
+    var b = void 0;
+    void ((b >>= 1) && (a = "FAIL"), b = 2);
+}
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/terser/compress/functions/loop_init_arg/output.js:5]: Effectively the same.

File: crates/swc_ecma_minifier/tests/terser/compress/identity/inline_identity_dont_lose_this_when_arg/output.js [hunk 10]
```diff
@@ -2,4 +2,5 @@
 const func_bag = {
     leak: leak
 };
-leak(func_bag.leak);
+var x;
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/terser/compress/identity/inline_identity_dont_lose_this_when_arg/output.js:5]: unused pass not enabled.

File: crates/swc_ecma_minifier/tests/terser/compress/identity/inline_identity_lose_this/output.js [hunk 11]
```diff
@@ -1,12 +1,14 @@
...
-func_bag.func2 = function () {
+func_bag.func2 = function() {
     return void 0 === this ? "PASS" : "FAIL";
 };
-console.log((0, func_bag.func)());
-console.log((0, func_bag.func2)());
+var x;
+console.log((x = func_bag.func)());
+var x1;
+console.log((x1 = func_bag.func2)());
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/terser/compress/identity/inline_identity_lose_this/output.js:14]: Same as above.

File: crates/swc_ecma_minifier/tests/terser/compress/drop_unused/issue_t161_top_retain_13/output.js [hunk 12]
```diff
@@ -1,3 +1,2 @@
-const g = ()=>y
-, y = 3;
-console.log(2, y, 4, 2 * y, 8, 4 * y, 2, g(), 4);
+const g = ()=>y, y = 3;
+console.log(2, y, 4, 2 * y, 8, 4 * y, 2, y, 4);
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/terser/compress/drop_unused/issue_t161_top_retain_13/output.js:2]: https://github.com/terser/terser/issues/1413

**Review by jridgewell — DISMISSED** [review]

**Review by kdy1 — DISMISSED** [review]
Thank you! CI is failing

---

swc-bump:
 - swc_ecma_minifier

**Comment by Austaras** [comment]
Yeah, TSC tests need update. It would be one thousand lines of update, mostly removing top level use strict.

**Comment by Austaras** [comment]
It's weird. I cannot reproduce the CI failure, seems like it's comparing to pre updated tests.

**Comment by kdy1** [comment]
I'll update them by myself. Thank you again for the work!

**Comment by kdy1** [comment]
I'll fix CI with another PR because I'm not sure about the way to reproduce test failure locally

**Pull request merged**